spatial/curve/hilbert_test.go

```diff
@@ -0,0 +1,372 @@
+		t.Run(fmt.Sprintf("Dimension/%d", dim), func(t *testing.T) {
+			for ord := 1; ord <= 4; ord++ {
+				t.Run(fmt.Sprintf("Order %d", ord), func(t *testing.T) {
+					testCurve(t, Hilbert4D{Order: ord})
+				})
+			}
+		})
+	}
+}
+
+func BenchmarkHilbert(b *testing.B) {
+	for N := 2; N <= 4; N++ {
+		b.Run(fmt.Sprintf("%dD/Curve", N), func(b *testing.B) {
+			for ord := 1; ord <= 10; ord++ {
+				b.Run(fmt.Sprintf("Order %d", ord), func(b *testing.B) {
+					h := newCurve(ord, N)
+					v := make([]int, N)
+					for i := range v {
+						v[i] = rand.Intn(1 << ord)
+					}
+					u := make([]int, N)
+					for n := 0; n < b.N; n++ {
+						copy(u, v)
+						h.Curve(u)
+					}
+				})
+			}
+		})
+
+		b.Run(fmt.Sprintf("%dD/Space", N), func(b *testing.B) {
+			for ord := 1; ord <= 10; ord++ {
+				b.Run(fmt.Sprintf("Order %d", ord), func(b *testing.B) {
+					h := newCurve(ord, N)
+					d := rand.Intn(1 << (ord * N))
+					for n := 0; n < b.N; n++ {
+						h.Space(d)
+					}
+				})
+			}
+		})
+	}
+}
+
+type curve interface {
+	Dims() []int
+	Len() int
+	Curve(v []int) int
+	Space(d int) []int
+}
+
+func newCurve(order, dim int) curve {
+	switch dim {
+	case 2:
+		return Hilbert2D{Order: order}
+	case 3:
+		return Hilbert3D{Order: order}
+	case 4:
+		return Hilbert4D{Order: order}
+	}
+	panic("invalid dimension")
+}
+
+func testCurve(t *testing.T, c curve) {
+	var errc int
+	fail := func() {
+		if errc < 10 {
+			errc++
+			t.Fail()
+		} else {
+			t.FailNow()
+		}
+	}
+
+	m := map[int][]int{}
+	curveRange(c, func(v []int) {
+		d := c.Curve(dup(v))
+		u := c.Space(d)
+		if !reflect.DeepEqual(v, u) {
+			t.Logf("Space is not the inverse of Curve for d=%d %v", d, v)
+			fail()
+		}
+
+		m[d] = dup(v)
+	})
+
+	D := 1
+	for _, v := range c.Dims() {
+		D *= v
+	}
+	for d := 0; d < D-1; d++ {
+		v, u := m[d], m[d+1]
+		if !adjacent(v, u) {
+			t.Logf("points %x and %x are not adjacent", d, d+1)
+			t.Logf("    %v -> %v", v, u)
+			fail()
+		}
+	}
+}
+
+func curveRange(c curve, fn func([]int)) {
+	size := c.Dims()
+	dimRange(len(size), size, make([]int, len(size)), fn)
+}
+
+func dimRange(dim int, size []int, v []int, fn func([]int)) {
+	if dim == 0 {
+		fn(v)
+		return
+	}
+
+	for i := 0; i < size[dim-1]; i++ {
+		v[dim-1] = i
+		dimRange(dim-1, size, v, fn)
+	}
+}
+
+// V may get mangled by Curve, so copy it
+func dup(v []int) []int {
+	u := make([]int, len(v))
+	copy(u, v)
+	return u
+}
+
+func adjacent(v, u []int) bool {
+	n := 0
+	for i := range v {
+		x := v[i] - u[i]
+		if x == 0 {
+			continue
+		} else if x > 1 || x < -1 {
+			return false
+		}
+		n++
+	}
+
+	return n == 1
+}
+
+func testCurveCase(t *testing.T, c curve, order int, expected []int) {
+	dim := len(c.Dims())
+	actual := make([]int, len(expected))
+	for i := range expected {
+		v := vec(i, order, dim)
+		actual[i] = c.Curve(dup(v))
+	}
+	if !reflect.DeepEqual(expected, actual) {
+		t.Logf("expected %v", expected)
+		t.Logf("got      %v", expected)
+		t.FailNow()
```

Comment:
This can be spelled with `t.Fatalf`.

Reply:
I did that so the two error messages would be aligned. Would you prefer `t.Fatalf("expected %v\ngot      %v", ...)`? I want them on separate lines, since they're arrays.

Also, the second one definitely should be using `actual`...

Reply:
The way that I approach this is the put a newline before the got and wanted lines; the testing logger automatically indents appropriately.

Like so, `t.Fatalf("unexpected result:\ngot: %v\nwant:%v", got, want)`.